### PR TITLE
TablePanel: Do not prefix columns with frame name if multiple frames and override active

### LINKED ISSUE #44971: DataFrame: Field name get's prefixed with frame name in Table panel after override is added
* Create table with two queries that return two NAMED frames

Screenshot from 2022-02-07 11-23-58

* Observe Table columns are just A and B
* Add an override rule

Column names now become FirstFrame A | First Frame B 

This is due to naming of frames are calculated during override process and cached, so even though Table panel should not ask for field names that need to account for other frames (as it's only showing a single frame) it get's a name that does account for multiple frames. 

https://github.com/grafana/grafana/blob/1680e284e5cc3f7e0a7ff747cbc007f966fce830/packages/grafana-data/src/field/fieldState.ts#L79

Screenshot from 2022-02-07 11-25-58

Not really sure how to fix this beyond clearing fieldstate.displayName after apply field overrides. But the best thing might be to just store a property in the fieldstate that the current cached fieldname was for "all frames" and should not be used for a single frame name.

### PR DESCRIPTION
**Pull request by mdvictor**
**What this PR does / why we need it**:
Fixes issue where some column headers would be prefixed with the frame name in the TablePanel when there would be multiple named frames and an override would be added.
After implementing this PR no column headers will be prefixed with the frame name.

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/grafana/issues/44971

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
think this is in the right direction but there should be no changes to Table panel, the issue is only in getFieldDisplayName and can be solved there. 

https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/field/fieldState.ts#L42

If you check there if the field state has multipleFrames but the value passed to getFieldDisplayName then you need to discard it and rebuild the name

**Comment by mdvictor** [comment]
I think I understand better now. Is this closer to what you envisioned?

File: packages/grafana-data/src/types/dataFrame.ts [hunk 1]
```diff
@@ -175,6 +175,11 @@ export interface FieldState {
    * @internal -- we will try to make this unnecessary
    */
   origin?: DataFrameFieldIndex;
+
+  /**
+   * Boolean value is true if field is in a larger data set with multiple frames.
+   */
+  multipleFrames?: boolean;
```
Comment (torkelo) [inline packages/grafana-data/src/types/dataFrame.ts:182]: This is not strictly true, as this only is set in calculateFieldDisplayName. Should add to the comment that this only related to the cached displayName

**Review by torkelo — APPROVED** [review]
Looks very close! added some things that think could improve the code

**Review by dprokop — APPROVED** [review]
LGTM!

**Comment by mdvictor** [comment]
Refactored the code as per suggestions!

File: packages/grafana-data/src/field/fieldState.ts [hunk 2]
```diff
@@ -38,14 +38,16 @@ export function getFrameDisplayName(frame: DataFrame, index?: number) {
 
 export function getFieldDisplayName(field: Field, frame?: DataFrame, allFrames?: DataFrame[]): string {
   const existingTitle = field.state?.displayName;
+  const calculatedForMultipleFrames = Boolean(field.state?.multipleFrames) && !Boolean(allFrames);
 
-  if (existingTitle) {
+  if (existingTitle && !calculatedForMultipleFrames) {
```
[the remark this answers is not in the document]
Comment (mdvictor) [reply packages/grafana-data/src/field/fieldState.ts:43]: I agree that the inversion makes it less understandable. Will try to get rid of it, but the equality `Boolean(field.state?.multipleFrames) == Boolean(allFrames?.length > 1)` brakes some tests and I also think it's not quite right.

We need to check when multipleFrames is true, which means allFrames was greater than 1 in a previous call on that field, but now allFrames is empty, which should trigger a recalculation.

File: packages/grafana-data/src/field/fieldState.ts [hunk 3]
```diff
@@ -38,9 +38,10 @@ export function getFrameDisplayName(frame: DataFrame, index?: number) {
 
 export function getFieldDisplayName(field: Field, frame?: DataFrame, allFrames?: DataFrame[]): string {
   const existingTitle = field.state?.displayName;
-  const calculatedForMultipleFrames = Boolean(field.state?.multipleFrames) && !Boolean(allFrames);
+  const canUseCachedDisplayName =
```
Comment (mdvictor) [inline packages/grafana-data/src/field/fieldState.ts:41]: I've modified the expression so it is no longer inverted, keeps tests working and respects the desired functionality.

File: packages/grafana-data/src/field/fieldState.ts [hunk 4]
```diff
@@ -38,14 +38,17 @@ export function getFrameDisplayName(frame: DataFrame, index?: number) {
 
 export function getFieldDisplayName(field: Field, frame?: DataFrame, allFrames?: DataFrame[]): string {
   const existingTitle = field.state?.displayName;
+  const canUseCachedDisplayName =
+    !Boolean(field.state?.multipleFrames) || (Boolean(field.state?.multipleFrames) && Boolean(allFrames));
```
Comment (torkelo) [inline packages/grafana-data/src/field/fieldState.ts:42]: Is not the same as `Boolean(field.state?.multipleFrames) ===  Boolean(allFrames && allFrames.length > 1)`

Comment (mdvictor) [reply packages/grafana-data/src/field/fieldState.ts:42]: The expression above breaks some tests. 
Also, maybe I don't understand correctly, but why should we allow the use of the cached display name if both multipleFrames and allFrames are either truthy/falsy?

**Pull request merged**